File: src/main/java/operator/CogroupOperator.java
package main.java.operator;

import java.util.Arrays;
import java.util.List;

import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaPairRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.api.java.function.VoidFunction;

import scala.Tuple2;

public class CogroupOperator {

	public static void main(String[] args) {
		SparkConf conf = new SparkConf().setAppName("ReduceOperator")
				.setMaster("local");
		JavaSparkContext sc = new JavaSparkContext(conf);
		
		List<Tuple2<String,String>> studentsList = Arrays.asList(
				new Tuple2<String,String>("1","xuruyun"),
				new Tuple2<String,String>("2","wangfei"),
				new Tuple2<String,String>("3","lixin"));
		
		/**
		 * 1 xuruyun 100
		 * 1 xuruyun 100
		 * 2 wangfei 90
		 * 2 wangfei 60
		 * 3 linxi 80
		 * 3 lixin 50
		 */
		List<Tuple2<String,String>> scoreList = Arrays.asList(
				new Tuple2<String,String>("1","100"),
				new Tuple2<String,String>("2","90"),
				new Tuple2<String,String>("3","80"),
				new Tuple2<String,String>("1","100"),
				new Tuple2<String,String>("2","60"),
				new Tuple2<String,String>("3","50"));
		
		JavaPairRDD<String,String> students = sc.parallelizePairs(studentsList);
		JavaPairRDD<String,String> scores = sc.parallelizePairs(scoreList);
		
		// cogroup 与 join不同！
		// 相当于，一个key join上所有value，都给放到一个Iterable里面去！
		JavaPairRDD<String, Tuple2<Iterable<String>, Iterable<String>>> studentScores = students.cogroup(scores);
		
		
		studentScores.foreach(new VoidFunction<Tuple2<String,Tuple2<Iterable<String>,Iterable<String>>>>() {
			
			private static final long serialVersionUID = 1L;

			@Override
			public void call(
					Tuple2<String, Tuple2<Iterable<String>, Iterable<String>>> tuple)
					throws Exception {
				System.out.println("student id : " + tuple._1);
				System.out.println("student name : " + tuple._2._1);
				System.out.println("student score : " + tuple._2._2);
			}
		});
		
		System.out.println(studentScores.collect());
		
		sc.close();
	}
}

File: src/main/scala/operator/CogroupOperator.scala
package main.scala.operator

import org.apache.spark.SparkConf
import org.apache.spark.SparkContext

/**
 * cogroup算子演示
 */
object CogroupOperator {
  def main(args: Array[String]): Unit = {
    val sparkConf = new SparkConf().setAppName("CogroupOperator")
    .setMaster("local")
    val sc = new SparkContext(sparkConf)
    
    val nameList = List((1,"wangbaoqiang"),(2,"marong"),(3,"songzhe"))
    val nameRdd = sc.parallelize(nameList)
    
    val scoreList = List((1,100),(2,59),(3,-1))
    val scoreRdd = sc.parallelize(scoreList)
    
    val gradeList = List((1,"三年级"),(2,"幼儿园"),(3,"五年级"))
    val gradeRdd = sc.parallelize(gradeList)
    
    val resultRdd = nameRdd.cogroup(scoreRdd,gradeRdd)
    
    val results = resultRdd.collect()
    for(result <- results){
      println(result)
      val rs  = result._2._3.seq
      for(r <- rs){
        println(r)
      }
    }
    
    sc.stop()
  }
}